lib/util/id3_utils.js

```diff
@@ -0,0 +1,298 @@
+    return size;
+  }
+
+  /**
+   * @param {Uint8Array} data
+   * @return {shaka.extern.MetadataRawFrame}
+   * @private
+   */
+  static getFrameData_(data) {
+    /*
+     * Frame ID       $xx xx xx xx (four characters)
+     * Size           $xx xx xx xx
+     * Flags          $xx xx
+     */
+    const type = String.fromCharCode(data[0], data[1], data[2], data[3]);
+    const size = shaka.util.Id3Utils.readSize_(data, 4);
+
+    // skip frame id, size, and flags
+    const offset = 10;
+
+    return {
+      type,
+      size,
+      data: data.subarray(offset, offset + size),
+    };
+  }
+
+  /**
+   * @param {shaka.extern.MetadataRawFrame} frame
+   * @return {?shaka.extern.MetadataFrame}
+   * @private
+   */
+  static decodeFrame_(frame) {
+    const BufferUtils = shaka.util.BufferUtils;
+    const StringUtils = shaka.util.StringUtils;
+
+    const metadataFrame = {
+      id: frame.type,
+      key: frame.type,
+      description: '',
+      data: '',
+      value: '',
+    };
+
+    if (frame.type === 'TXXX') {
+      /*
+       * Format:
+       * [0]   = {Text Encoding}
+       * [1-?] = {Description}\0{Value}
+       */
+      if (frame.size < 2) {
+        return null;
+      }
+      if (frame.data[0] !== shaka.util.Id3Utils.UTF8_encoding) {
+        shaka.log.warning('Ignore frame with unrecognized character ' +
+            'encoding');
+        return null;
+      }
+      const descriptionEndIndex = frame.data.subarray(1).indexOf(0);
+
+      if (descriptionEndIndex === -1) {
+        return null;
+      }
+      const description = StringUtils.fromUTF8(
+          BufferUtils.toUint8(frame.data, 1, descriptionEndIndex));
+      const data = StringUtils.fromUTF8(
+          BufferUtils.toUint8(frame.data, 2 + descriptionEndIndex));
+
+      metadataFrame.description = description;
+      metadataFrame.data = data;
+      metadataFrame.value = metadataFrame.data;
+      return metadataFrame;
+    } else if (frame.type === 'WXXX') {
+      /*
+       * Format:
+       * [0]   = {Text Encoding}
+       * [1-?] = {Description}\0{URL}
+       */
+      if (frame.size < 2) {
+        return null;
+      }
+      if (frame.data[0] !== shaka.util.Id3Utils.UTF8_encoding) {
+        shaka.log.warning('Ignore frame with unrecognized character ' +
+            'encoding');
+        return null;
+      }
+      const descriptionEndIndex = frame.data.subarray(1).indexOf(0);
+
+      if (descriptionEndIndex === -1) {
+        return null;
+      }
+      const description = StringUtils.fromUTF8(
+          BufferUtils.toUint8(frame.data, 1, descriptionEndIndex));
+      const data = StringUtils.fromUTF8(
+          BufferUtils.toUint8(frame.data, description.length + 1));
+
+      metadataFrame.description = description;
+      metadataFrame.data = data;
+      metadataFrame.value = metadataFrame.data;
+      return metadataFrame;
+    } else if (frame.type === 'PRIV') {
+      /*
+       * Format: <text string>\0<binary data>
```

Comment:
You call the nul byte location "ownerEndIndex".  Should this format say "owner" instead of "text string"?